From 7425fb96afafeb6a8457e85c49eba86d36a6cebe Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:02:38 -0400
Subject: [usb] wusbcore: preserve endianness of cached descriptors

Message-id: <1410887463-20674-87-git-send-email-dzickus@redhat.com>
Patchwork-id: 95508
O-Subject: [RHEL7 PATCH 086/591] usb: wusbcore: preserve endianness of cached descriptors
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 1653d2f88f43f6780eace1faeeb7b7adde8c10fb
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Mon Oct 7 10:07:51 2013 -0500

    usb: wusbcore: preserve endianness of cached descriptors

    Do not overwrite the multi-byte fields of usb_wa_descriptor with their
    cpu format values after reading the descriptor.  Leave the values as
    __le16 and swap on use.  This is more consistent with other uses of USB
    descriptors.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/hwa-hc.c b/drivers/usb/host/hwa-hc.c
index e58b924..e5fb3cf 100644
--- a/drivers/usb/host/hwa-hc.c
+++ b/drivers/usb/host/hwa-hc.c
@@ -570,14 +570,10 @@ found:
   goto error;
  }
  wa->wa_descr = wa_descr = (struct usb_wa_descriptor *) hdr;
- /* Make LE fields CPU order */
- wa_descr->bcdWAVersion = le16_to_cpu(wa_descr->bcdWAVersion);
- wa_descr->wNumRPipes = le16_to_cpu(wa_descr->wNumRPipes);
- wa_descr->wRPipeMaxBlock = le16_to_cpu(wa_descr->wRPipeMaxBlock);
- if (wa_descr->bcdWAVersion > 0x0100)
+ if (le16_to_cpu(wa_descr->bcdWAVersion) > 0x0100)
   dev_warn(dev, "Wire Adapter v%d.%d newer than groked v1.0\n",
-    wa_descr->bcdWAVersion & 0xff00 >> 8,
-    wa_descr->bcdWAVersion & 0x00ff);
+    le16_to_cpu(wa_descr->bcdWAVersion) & 0xff00 >> 8,
+    le16_to_cpu(wa_descr->bcdWAVersion) & 0x00ff);
  result = 0;
 error:
  return result;
diff --git a/drivers/usb/wusbcore/wa-rpipe.c b/drivers/usb/wusbcore/wa-rpipe.c
index fd4f1ce..e3b913b 100644
--- a/drivers/usb/wusbcore/wa-rpipe.c
+++ b/drivers/usb/wusbcore/wa-rpipe.c
@@ -477,7 +477,7 @@ error:
  */
 int wa_rpipes_create(struct wahc *wa)
 {
- wa->rpipes = wa->wa_descr->wNumRPipes;
+ wa->rpipes = le16_to_cpu(wa->wa_descr->wNumRPipes);
  wa->rpipe_bm = kzalloc(BITS_TO_LONGS(wa->rpipes)*sizeof(unsigned long),
           GFP_KERNEL);
  if (wa->rpipe_bm == NULL)
diff --git a/include/linux/usb/wusb-wa.h b/include/linux/usb/wusb-wa.h
index 9ae7e29..c125713 100644
--- a/include/linux/usb/wusb-wa.h
+++ b/include/linux/usb/wusb-wa.h
@@ -279,11 +279,11 @@ struct wa_xfer_result {
 struct usb_wa_descriptor {
  u8 bLength;
  u8 bDescriptorType;
- u16 bcdWAVersion;
+ __le16 bcdWAVersion;
  u8 bNumPorts;  /* don't use!! */
  u8 bmAttributes;  /* Reserved == 0 */
- u16 wNumRPipes;
- u16 wRPipeMaxBlock;
+ __le16 wNumRPipes;
+ __le16 wRPipeMaxBlock;
  u8 bRPipeBlockSize;
  u8 bPwrOn2PwrGood;
  u8 bNumMMCIEs;
-- 
1.7.1